### ISSUE TITLE
Native support for bHaptics and DIY haptics

### ISSUE THREAD

**Issue by sacred0** [vrchat-community/osc#158]
**What's the idea?**
A solution, either as a part of the Avatar descriptor or as custom components, to allow for OSC (or direct) output to haptics devices. 
This could be accomplished in multiple ways, such as "haptic contacts", which wouldn't count towards the performance rank due to only being local. They could be set up similarly to the current way contacts are set up in the avatar descriptor, or as components that are manually placed. The components could either be spheres/capsules like the current contacts, or "collision meshes" that can be formed to the body shape and would map to a configured set of haptic nodes.

Alternatively, a system similar to the existing controller haptics could be used, where collision with the avatar mesh itself could be used for output. I'm not too sure how this could be easily mapped to haptics gear, but I'm sure there's a way.

**Is there another way?**
As a stop-gap solution, making Local Only contacts not count towards the performance rank would mean that the current systems would still remain functional, and fix the issue of being forced into Very Poor. 

**Additional context**
Currently, a ridiculous amount of contacts is needed for adding haptics support, which immediately drives the avatar into Very Poor with no other good alternative. Optimisation is very important to me when I work on avatars, and not being able to have a better performance rank because of this is quite demoralizing. The current systems also only provide binary interaction, meaning there is no difference between a light touch and getting punched. This could in theory be improved using proximity contacts and more advanced software, but this currently doesn't exist.

**Comment by koshisan** [comment]
"As a stop-gap solution, making Local Only contacts not count towards the performance rank would mean that the current systems would still remain functional, and fix the issue of being forced into Very Poor."

This is so important, would take basically no effort at all and make life easier for virtually anyone trying to add hardware. I really wish this would gain a bit more traction...

(actually, one could even consider local-only contacts counting towards the performance rating as a bug in the first place...)

**Comment by sacred0** [comment]
> "As a stop-gap solution, making Local Only contacts not count towards the performance rank would mean that the current systems would still remain functional, and fix the issue of being forced into Very Poor."
> 
> This is so important, would take basically no effort at all and make life easier for virtually anyone trying to add hardware. I really wish this would gain a bit more traction...
> 
> (actually, one could even consider local-only contacts counting towards the performance rating as a bug in the first place...)

This has actually been implemented fairly recently, and is live ^^ Local only contacts no longer count towards perf rank, which is both useful for haptics and for some avatar systems.

**Comment by koshisan** [comment]
Yep, I just noticed yesterday! I am still in testing because the documentation isn't 100% clear if contacts declared as local only can be triggered by other users (which would of course be the main requirement for haptics). I am especially confused because the bHaptics OSC integration, which also was updated in late December, does not configure the points as local...
But if they indeed work that way, this would actually have been my favourite christmas present this year!

**Comment by sacred0** [comment]
They can, because from your point of view, the local-only contacts exist and so do the senders on other peoples' avatars. 

The official bHaptics prefabs are a little out of date. There is this more up to date one by Buddy that includes a prefab for the vest, that is also easier to set up because of VRCFury. https://github.com/JustBuddy/simple-bHaptics_VRC. I believe it should have the contacts correctly marked as local-only.

**Comment by koshisan** [comment]
Thanks for the clarification and the link!
